### ISSUE TITLE
AttributeError: 'module' object has no attribute 'argv'

### ISSUE THREAD

**Issue by pavelschon** [pybind/pybind11#2061]
## Issue description

`sys.argv` is not available when using embedded interpreter.

## Reproducible example code

```c++
#include <pybind11/embed.h>

namespace py = pybind11;

int main( int argc, char** argv )
{
    py::scoped_interpreter guard{};

    py::print(py::module::import("sys").attr("argv"));
}

```

```
$ ./test
terminate called after throwing an instance of 'pybind11::error_already_set'
  what():  AttributeError: 'module' object has no attribute 'argv'
Aborted
```

## Workaround
Workaround is possible, however, it requires to call python C-API function and linkage to libpython.so.

```c++
#include <pybind11/embed.h>

namespace py = pybind11;

int main( int argc, char** argv )
{
    py::scoped_interpreter guard{};

    PySys_SetArgv(argc, argv);

    py::print(py::module::import("sys").attr("argv"));
}
```
```
$ ./test
['./test']
```
I feel that pybind11 should provide API to set _argc+argv_ without need to call Python C-API.
Related issue: #1918

**Comment by pavelschon** [comment]
Just an idea: pass argc, argv to constructor of scoped_interpreter:
```c++
py::scoped_interpreter guard{argc, argv};
```

**Comment by bstaletic** [comment]
I can't reproduce this. Can you recheck again?

**Comment by YannickJadoul** [comment]
Also related to #1768 (though there `argv` exists, but is empty.

**Comment by YannickJadoul** [comment]
It might deserve a note in the docs, at least. But an extra constructor for `guard` doesn't seem wrong either, to me.

**Comment by YannickJadoul** [comment]
> It might deserve a note in the docs, at least. But an extra constructor for `scoped_interpreter` doesn't seem wrong either, to me.

**Comment by bstaletic** [comment]
@YannickJadoul I'd like to make sure there is a problem that needs solving first. If it works, like it worked for me, I don't think we should add another constructor "just because".

**Comment by pavelschon** [comment]
> @YannickJadoul I'd like to make sure there is a problem that needs solving first. If it works, like it worked for me, I don't think we should add another constructor "just because".

Yes, there is a problem: sys.argv is not set by pybind11 automatically as one may expect.
Workaround is to explicitly call  PySys_SetArgv(argc, argv);.

I proposed a solution: add alternate constructor to py::scoped_interpreter accepting argc+argv.

**Comment by YannickJadoul** [comment]
> Yes, there is a problem: sys.argv is not set by pybind11 automatically as one may expect.

Well, we could also claim/decide that expectation is wrong. If you never pass `argv` and `argc`, it's a bit of a leap they would be set.
Also, you're embedding the Python interpreter potentially in the middle of your program. To what `argv` and `argc` would you expect things to be set?

**Comment by YannickJadoul** [comment]
Also, what version of Python etc are you using, because I cannot reproduce:

```
$ ./test
['']
```

**Comment by pavelschon** [comment]
@YannickJadoul 
_Also, what version of Python etc are you using_
We're on python 2.7 on RHEL 7.
FYI Python 2.7 is unsupported as upstream project, but is still supported within RHEL ecosystem.


 _...because I cannot reproduce:_

```
$ ./test
['']
```
If you're example does not crash, but instead prints [''] which is also obviously wrong. First arguments should be name of the script/executable like this:
```
$ ./test
['./test']
```

**Comment by YannickJadoul** [comment]
> If your example does not crash, but instead prints [''] which is also obviously wrong.

What is obviously wrong about this?

> First argument should be name of the script/executable like this:
> 
> ```
> $ ./test
> ['./test']
> ```

I strongly disagree. In your example case, where from `main` you immediately set up an interpreter, that would make sense, yes.
But in general, no, because a Python interpreter could be started deep down in a project, where the `argv` and `argc` from the program's `main` do not make sense. Consider for example using Python as a custom shell within a GUI application, or within an IDE: I don't see how the arguments to the GUI application make sense to be set inside that Python shell.
You're not starting Python from the command line with some arguments, you've started a program embedding Python from the command line with some arguments.

**Comment by bstaletic** [comment]
> First argument should be name of the script/executable like this:

Are you sure about that? Do the following: `python -c 'print(__import__("sys").argv)'`. Python's `argv` doesn't contain the name of the executable.

**Comment by YannickJadoul** [comment]
> Are you sure about that? Do the following: `python -c 'print(__import__("sys").argv)'`. Python's `argv` doesn't contain the name of the executable.

True. And that's even a moot case, for embedding. That's calling _Python_ from the command line. That's not using Python in another program that's called from the command line.

**Comment by bstaletic** [comment]
One more thing.

        PySys_SetArgv(argc, argv);

If `argc` is `int` and `argv` is `char**`, this doesn't even compile, since `PySys_SetArgv` on python3 takes an `int` and, believe it or not, `wchar_t**`.

**Comment by kotbegemot** [comment]
One of the possible solutions.

```C++
py::list argv;
py::module::import("sys").add_object("argv", argv);
```
We use this approach in project rocketjoe https://github.com/cyberduckninja/RocketJoe/blob/master/components/python_sandbox/source/python_sandbox.cpp#L310.

**Comment by YannickJadoul** [comment]
> ```
> 
> ```
> 
> 
> If `argc` is `int` and `argv` is `char**`, this doesn't even compile, since `PySys_SetArgv` on python3 takes an `int` and, believe it or not, `wchar_t**`.

Ugh.

https://docs.python.org/3/c-api/init.html#c.PySys_SetArgv

> Use Py_DecodeLocale() to decode a bytes string to get a wchar_* string.


> One of the possible solutions.

Anyway, yes, there are lots of solutions. My point is mainly that just passing on `argc` and `argv` is not the obvious default and that it requires tought and consideration from the programmer/designer anyway.

**Comment by pavelschon** [comment]
> > First argument should be name of the script/executable like this:
> 
> Are you sure about that? Do the following: `python -c 'print(__import__("sys").argv)'`. Python's `argv` doesn't contain the name of the executable.

Well, this does:
```
$ cat test.py
#!/usr/bin/env python3.5

import sys

print(sys.argv)
```
```
$ ./test.py 
['./test.py']
```
The same is printed with python2.7.

The idea, which allows complete control over what goes into argv - programmer could pass anything (or nothing) to the constructor of scoped_interpreter:
```
#include <pybind11/embed.h>

namespace py = pybind11;

int main( int argc, char** argv )
{
    py::scoped_interpreter guard{argc, argv};

    py::print(py::module::import("sys").attr("argv"));
}
```

**Comment by YannickJadoul** [comment]
> Well, this does:

That's beside the point. The point is that @bstaletic showed that there are situations where Python itself doesn't set `sys.argv` either.

> The idea, which allows complete control over what goes into argv - programmer could pass anything (or nothing) to the constructor of scoped_interpreter:

Now you have complete control over `argv` and `argc`, with `PySys_SetArgv`. Or with @kotbegemot's suggestion of `py::module::import("sys").attr("argv") = argv;`.

Just to be clear, I don't per se object to this extra constructor, but I don't see how it would allow you to do anything you can't do now. And I do object to your strong claim that
> If your example does not crash, but instead prints [''] which is also obviously wrong

So maybe just a note in the docs would be just as good as the extra constructor, because it wouldn't add anything.

**Comment by pavelschon** [comment]
I'm fine if my idea is not implemented. I just saw a relationship between instantiating the _scoped_interpreter_ and setting up _argc+argv_.
I like the idea of @kotbegemot - let's close this topic.

**Comment by YannickJadoul** [comment]
Alright, feel free to open a PR, if you want! We're currently cleaning up the repo, but in a few weeks or months, we should be able to reasonably quickly judge newly incoming PRs.

**Comment by pavelschon** [comment]
Implemented with #2341
